### PR TITLE
(FACT-1361) Create networking fact acceptance test

### PR DESCRIPTION
**Pull request by phongdly**
Prior to this PR, networking facts were verified in each individual test cases in acceptance/tests/facts/<platform>.rb

This PR pulls out the networking portion of all the tests in acceptance/tests/facts into new acceptance test, acceptance/tests/facts/networking_facts.rb

This will also help to test networking facts in many more platforms in puppet-agent CI, instead of being confined to only 9 platforms in acceptance/tests/facts

### CODE REVIEW

File: acceptance/tests/facts/networking_facts.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
+test_name 'FACT-1361 - C59029 networking facts should be fully populated'
```
Comment (er0ck) [inline acceptance/tests/facts/networking_facts.rb]: test_name and step should have `do/end` to create blocks, and proper indentation in logs

File: acceptance/tests/facts/networking_facts.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+#
+# This test is intended to ensure that networking facts resolves
+# as expected in AIO across supported platforms.
+#
+
+skip_test "Networking fact test is confined to AIO" if @options[:type] != 'aio'
```
Comment (er0ck) [inline acceptance/tests/facts/networking_facts.rb]: this is targeted at master, so i don't think this is required.  even if merged down to stable, it'll still always be aio.  unless i'm missing something about network devices, where they aren't run with aio type?

File: acceptance/tests/facts/networking_facts.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+@ip_regex = /^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/
+
+expected_networking = {
+    "networking.dhcp"     => @ip_regex,
+    "networking.ip"       => @ip_regex,
+    "networking.ip6"      => /[a-f0-9]+:+/,
+    "networking.mac"      => /[a-f0-9]{2}:/,
+    "networking.mtu"      => /\d+/,
+    "networking.netmask"  => /\d+\.\d+\.\d+\.\d+/,
```
Comment (er0ck) [inline acceptance/tests/facts/networking_facts.rb]: i think each of these octets can be restricted to... octets?  i.e.: this will always be 1-3 digits per octet, AFAIK.
@branan are these never in the `192.168.0.0/24` type format?

File: acceptance/tests/facts/networking_facts.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    "networking.netmask6" => /[a-f0-9]+:/,
```
Comment (er0ck) [inline acceptance/tests/facts/networking_facts.rb]: i think these are always `2001:0db8:0123:4567:89ab:cdef:1234:5678`  
as in 8 sets of 4 hex encoded chars separated by colons

File: acceptance/tests/facts/networking_facts.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+agents.each do |agent|
+
+  step "Ensure a primary networking interface was determined."
+  primary_interface = fact_on(agent, 'networking.primary')
+  refute_empty(primary_interface)
+
+  expected_bindings = {
+      "\"networking.interfaces.#{primary_interface}.bindings.0.address\"" => /\d+\.\d+\.\d+\.\d+/,
+      "\"networking.interfaces.#{primary_interface}.bindings.0.netmask\"" => /\d+\.\d+\.\d+\.\d+/,
+      "\"networking.interfaces.#{primary_interface}.bindings.0.network\"" => /\d+\.\d+\.\d+\.\d+/,
```
Comment (er0ck) [inline acceptance/tests/facts/networking_facts.rb]: can we re-use the netmask and ip regexes here?

File: acceptance/tests/facts/networking_facts.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      "\"networking.interfaces.#{primary_interface}.bindings6.0.address\"" => /[a-f0-9:]+/,
+      "\"networking.interfaces.#{primary_interface}.bindings6.0.netmask\"" => /[a-f0-9:]+/,
+      "\"networking.interfaces.#{primary_interface}.bindings6.0.network\"" => /[a-f0-9:]+/
+  }
+
+  case agent['platform']
+    when /solaris/, /eos/
+      #remove the invalid networking facts on aix and solaris
```
Comment (er0ck) [inline acceptance/tests/facts/networking_facts.rb]: comment appears incorrect

File: acceptance/tests/facts/networking_facts.rb (hunk 4 above)
Comment (seangriff) [inline acceptance/tests/facts/networking_facts.rb]: The hex characters do come in groups for four, and there are 8 groups, but there is a short hand that allows skipping one run of contiguous 0s.  An address may look like 2001:0db8::1234:5678, to mean 2001:0db8:0000:0000:0000:0000:1234:5678 (zeros between the two contiguous colons).  This makes a recognizing regexp very nasty (see http://stackoverflow.com/questions/53497/regular-expression-that-matches-valid-ipv6-addresses)

File: acceptance/tests/facts/networking_facts.rb (hunk 6 above)
Comment (seangriff) [inline acceptance/tests/facts/networking_facts.rb]: Nice work Phong.  This is much better than before.

File: acceptance/tests/facts/networking_facts.rb (hunk 4 above)
Comment (er0ck) [inline acceptance/tests/facts/networking_facts.rb]: hmmm, there is also:
http://ruby-doc.org/stdlib-2.0.0/libdoc/ipaddr/rdoc/IPAddr.html#method-i-ip6_int

TL;DR: use ruby's networking code to check them

File: acceptance/tests/facts/networking_facts.rb (hunk 1 above)
Comment (phongdly) [inline acceptance/tests/facts/networking_facts.rb]: will add do/end for test_name in the next push

File: acceptance/tests/facts/networking_facts.rb (hunk 2 above)
Comment (phongdly) [inline acceptance/tests/facts/networking_facts.rb]: will remove the lind in the next push

File: acceptance/tests/facts/networking_facts.rb (hunk 3 above)
Comment (phongdly) [inline acceptance/tests/facts/networking_facts.rb]: That makes sense, netmask should be restricted to octets. 
Netmask type A format: 255.0.0.0
Netmask type B format: 255.255.0.0
Netmask type C format: 255.255.255.0

But in some customized netmask, any 255 in any octet can be replaced by either 128, 193, 224, 240, 248, 252, or 254.

So I will add the @netmask_regex = /^(((128|192|224|240|248|252|254)\.0\.0\.0)|(255\.(0|128|192|224|240|248|252|254)\.0\.0)|(255\.255\.(0|128|192|224|240|248|252|254)\.0)|(255\.255\.255\.(0|128|192|224|240|248|252|254)))$/ in the next push

Comment (er0ck) [inline acceptance/tests/facts/networking_facts.rb]: that seems a bit hard to read.
i'd probably just do:
`/\d{1,3}?\.\d{1,3}?\.\d{1,3}?\.\d{1,3}?\./`

File: acceptance/tests/facts/networking_facts.rb [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        expected_bindings.delete("\"networking.interfaces.#{primary_interface}.bindings6.0.netmask\"")
+        expected_bindings.delete("\"networking.interfaces.#{primary_interface}.bindings6.0.network\"")
+
+      when /sles/
+        #some sles VMs do not have networking.dhcp
+        expected_networking.delete("networking.dhcp")
+
+    end
+
+    step "Ensure the Networking fact resolves with reasonable values for at least one interface"
```
Comment (er0ck) [inline acceptance/tests/facts/networking_facts.rb]: `step` needs do/end blocks also

**Comment by er0ck** [comment]
looks good.  please add do/end blocks to all `step`'s and i can merge.

File: acceptance/tests/facts/networking_facts.rb (hunk 7 above)
Comment (er0ck) [inline acceptance/tests/facts/networking_facts.rb]: @phongdly please add do/end blocks around your steps
